Using function argument as `new` finalizer fails to compile

`new` with finalizer works well if I pass a global function as the finilizer, however if I pass an argument of a function (which is of course also a function) to it compilation will fail during calling `gcc`.


### Example
<!-- Paste your example in the code-block below. -->
```nim
type
  A = ref object

proc callback(a: A) {. nimcall .} = 
  discard

proc foo(callback: proc(a: A) {. nimcall .}) =
  var a: A
  new(a, callback)

foo(callback)
```


### Current Output
```
Hint: used config file '/home/wtli/.choosenim/toolchains/nim-0.19.0/config/nim.cfg' [Conf]
Hint: system [Processing]
Hint: test_callback [Processing]
CC: test_callback
Error: execution of an external compiler program 'gcc -c  -w  -I/home/wtli/.choosenim/toolchains/nim-0.19.0/lib -o /home/wtli/.cache/nim/test_callback_d/test_callback.c.o /home/wtli/.cache/nim/test_callback_d/test_callback.c' failed with exit code: 1

/home/wtli/.cache/nim/test_callback_d/test_callback.c: In function ‘test_callbackDatInit000’:
/home/wtli/.cache/nim/test_callback_d/test_callback.c:178:54: error: ‘callback’ undeclared (first use in this function)
 (&NTI_DNgur79be8Ev9cbhq3iBkzxA_)->finalizer = (void*)callback;
                                                      ^
/home/wtli/.cache/nim/test_callback_d/test_callback.c:178:54: note: each undeclared identifier is reported only once for each function it appears in
```


### Expected Output
Successful compilation, or an error message showing I can only use global functions as finalizer.



### Additional Information
```
Nim Compiler Version 0.19.0 [Linux: amd64]
Compiled at 2018-09-26
Copyright (c) 2006-2018 by Andreas Rumpf

git hash: f6c5c636bb1a1f4e1301ae0ba5a8afecef439132
active boot switches: -d:release
```

--- a/compiler/semmagic.nim
+++ b/compiler/semmagic.nim
@@ -410,4 +410,9 @@ proc magicsAfterOverloadResolution(c: PContext, n: PNode,
       result = n
     else:
       result = plugin(c, n)
+  of mNewFinalize:
+    # Make sure the finalizer procedure refers to a procedure
+    if n[^1].kind == nkSym and n[^1].sym.kind notin {skProc, skFunc}:
+      localError(c.config, n.info, "finalizer must be a direct reference to a procedure")
+    result = n
   else: result = n

--- a/tests/errmsgs/t10376.nim
+++ b/tests/errmsgs/t10376.nim
@@ -1,0 +1,31 @@
+discard """
+  errormsg: "finalizer must be a direct reference to a procedure"
+  line: 29
+"""
+
+type
+  A = ref object
+
+proc my_callback(a: A) {. nimcall .} =
+  discard
+
+proc foo(callback: proc(a: A) {. nimcall .}) =
+  var x1: A
+  new(x1, proc (x: A) {.nimcall.} = discard)
+  var x2: A
+  new(x2, func (x: A) {.nimcall.} = discard)
+
+  var x3: A
+  proc foo1(a: A) {.nimcall.} = discard
+  new(x3, foo1)
+  var x4: A
+  func foo2(a: A) {.nimcall.} = discard
+  new(x4, foo2)
+
+  var x5: A
+  new(x5, my_callback)
+
+  var x6: A
+  new(x6, callback)
+
+foo(my_callback)

